### PR TITLE
Serialize Outbound Message on Flush

### PR DESCRIPTION
**Pull request by original-brownbear**
Follow up to #56961:

We can be a little more efficient than just serializing at the IO loop by serializing
only when we flush to a channel. This has the advantage that we don't serialize a long
queue of messages for a channel that isn't writable for a longer period of time (unstable network,
actually writing large volumes of data, etc.).
Also, this further reduces the time for which we hold on to the write buffer for a message,
making allocations because of an empty page cache recycler pool less likely.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java [hunk 1]
```diff
@@ -190,13 +190,23 @@ private void failQueuedWrites() {
...
+        WriteOperation(OutboundHandler.SendContext context, ChannelPromise promise) {
+            this.context = context;
             this.promise = promise;
         }
+
+        ByteBuf buffer() throws IOException {
+            if (buf == null) {
+                buf = Netty4Utils.toByteBuf(context.get());
+                context = null;
+            }
```
Comment (DaveCTurner) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java:208]: Can we `assert context == null` here?

**Review by DaveCTurner — APPROVED** [review]
LGTM assuming that there's only a single thread in play here; I think there is but I'm not 100% familiar with this code.

Requested an assertion that I think would catch any bad threading anyway.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java (hunk 1 above)
Comment (original-brownbear) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java:208]: Sure thing, added it. But yes, all of this is happening on a single IO thread thread so this is all safe I think :)

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**